Detectors/MUON/MID/Raw/src/RawFileReader.cxx

```diff
@@ -0,0 +1,170 @@
+{
+  /// Clears the bytes and counters
+  mBytes.clear();
+  mHBCounters.fill(0);
+}
+
+template <typename T>
+bool RawFileReader<T>::hasFullInfo()
+{
+  /// Tests if we have read the same number of HBs for all GBT links
+
+  // We assume here that we read the data of one HV for all of the GBT links
+  // before moving to the next HB
+  for (uint16_t feeId = 1; feeId < crateparams::sNGBTs; ++feeId) {
+    if (mHBCounters[feeId] != mHBCounters[0]) {
+      return false;
+    }
+  }
+  return mHBCounters[0] != 0;
+}
+
+template <typename T>
+bool RawFileReader<T>::readAllGBTs(bool reset)
+{
+  /// Keeps reading the file until it reads the information of all GBT links
+  /// It returns the number of HBs read
+
+  if (reset) {
+    clear();
+  }
+
+  while (!hasFullInfo()) {
+    if (!readHB()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+template <typename T>
+bool RawFileReader<T>::replaceRDH(size_t headerIndex)
+{
+  /// Replaces the current RDH with a custom one if needed.
+  /// This is done to be able to correctly read test data
+  /// that have a wrong RDH
+  if (mCustomRDH.offsetToNext > 0) {
+    header::RAWDataHeader* rdh = reinterpret_cast<header::RAWDataHeader*>(&mBytes[headerIndex]);
+    *rdh = mCustomRDH;
+    return true;
+  }
+  return false;
+}
+
+template <typename T>
+bool RawFileReader<T>::readHB(bool sendCompleteHBs)
+{
+  /// Reads one HB
+  if (mState != 0) {
+    return false;
+  }
+  auto gbtId = 0;
+  bool isHBClosed = false;
+  while (!isHBClosed) {
+    // Read header
+    size_t headerIndex = mBytes.size();
+    read(raw::sHeaderSizeInBytes);
+
+    // The check on the eof needs to be placed here and not at the beginning of the function.
+    // The reason is that the eof flag is set if we try to read after the eof
+    // But, since we know the size, we read up to the last character.
+    // So we turn on the eof flag only if we try to read past the last data.
+    // Of course, we resized the mBytes before trying to read.
+    // Since we read 0, we need to remove the last bytes
+    if (mFile.eof()) {
+      mBytes.resize(headerIndex);
+      if (mReadContinuous) {
+        mFile.clear();
+        mFile.seekg(0, std::ios::beg);
+        read(raw::sHeaderSizeInBytes);
+      } else {
+        mState = 1;
+        return false;
+      }
+    }
+    replaceRDH(headerIndex);
+    // We use the buffer only to correctly initialize the RDH
+    mBuffer.nextHeader(mBytes);
+    isHBClosed = mBuffer.isHBClosed();
+    gbtId = mBuffer.getRDH()->feeId;
+    // CAVEAT: do not call mBuffer.getRDH() beyond this line
+    // To save memory/CPU time, the RawBuffer does not hold a copy of the buffer,
+    // but just a span of it.
+    // If we add bytes to mBytes, the vector can go beyond the capacity
+    // and the memory is re-allocated.
+    // If this happens, the span is no longer valid, and we can no longer
+    // call mBuffer.getRDH() until we pass it mBytes again
+    if (gbtId >= crateparams::sNGBTs) {
```

Comment:
Did you consider ways to unit-test your decoder ? So this FIXME won't be forgotten in the future ?

Reply:
This is not an issue with my decoder, but an issue with the test files.
The problem is that the feeId in the test files is currently simply wrong, so I cannot rely on it. Once this will be fixed in the test files, it will be automatically ok even here. But there is not much I can do about it....